mitiq/pec/representations/learning.py

```diff
@@ -15,16 +15,109 @@
+    fraction_non_clifford: float = 0.2,
+    training_random_state: np.random.RandomState = None,
+    epsilon0: float = 0.05,
+    observable: Optional[Observable] = None,
+    **minimize_kwargs: Dict["str", Any],
+) -> Tuple[bool, float]:
+    r"""This function learns the depolarizing noise parameter (epsilon)
+    associated to a set of input operations. The learning process is based on
+    the execution of a set of training circuits on a noisy backend and on a
+    classical simulator. The training circuits are near-Clifford approximations
+    of the input circuit. A depolarizing noise model characterization is
+    assumed.
+
+    Args:
+        operations_to_learn: The ideal operations to learn the noise model of.
+        circuit: The full quantum program as defined by the user.
+        ideal_executor: Simulates a circuit and returns
+        a noiseless `QuantumResult`.
+        noisy_executor: Executes a circuit on a noisy backend
+        and returns a `QuantumResult`.
+        num_training_circuits: Number of near-Clifford circuits to be
+            generated for training.
+        epsilon0: Initial guess for noise strength.
+        training_random_state: Random seed for generating training circuits.
+        observable (optional): Observable to compute the expectation value of.
+            If None, the `executor` must return an expectation value. Otherwise
+            the `QuantumResult` returned by `executor` is used to compute the
+            expectation of the observable.
+
+    Returns:
+        A flag indicating whether or not the optimizer exited successfully and
+        the optimized noise strength epsilon.
+    """
+    training_circuits = generate_training_circuits(
+        circuit=circuit,
+        num_training_circuits=num_training_circuits,
+        fraction_non_clifford=fraction_non_clifford,
+        random_state=training_random_state,
+    )
+
+    ideal_values = np.array(
+        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+    )
+
+    def depolarizing_noise_loss_function(
+        epsilon: List[int],
+        operations_to_mitigate: List[QPROGRAM],
+        training_circuits: List[QPROGRAM],
+        ideal_values: np.ndarray,
+        noisy_executor: Executor,
+        pec_kwargs: Dict["str", Any],
+        observable: Optional[Observable] = None,
+    ) -> float:
+        return biased_noise_loss_function(
+            np.array([epsilon[0], 0]),
+            operations_to_mitigate,
+            training_circuits,
+            ideal_values,
+            noisy_executor,
+            pec_kwargs,
+            observable,
+        )
+
+    result = minimize(
+        depolarizing_noise_loss_function,
+        epsilon0,
+        args=(
+            operations_to_learn,
+            training_circuits,
+            ideal_values,
+            noisy_executor,
+            pec_kwargs,
+            observable,
+        ),
+        method="Nelder-Mead",
+        **minimize_kwargs,
```

Comment:
I added an argument `solver` to `learn_depolarizing_noise_parameter` with `"Nelder-Mead"` as default, since Nelder-Mead works reliably for the tests, unlike BFGS and others.